src/Utility/AnaSol.C

```diff
@@ -17,13 +17,54 @@
 #include "Utilities.h"
 #include "IFEM.h"
 #include "tinyxml.h"
+
+#include <autodiff/reverse/var.hpp>
+
 #ifdef HAS_HDF5
-#include "ProcessAdm.h"
-#include "HDF5Writer.h"
 #include "FieldFunctions.h"
 #endif
 
 
+namespace {
+
+//! \brief Function returning the gradient of a scalar or vector function.
+template<class FromFunc, class ToFunc, class Ret>
+class GradientFunc : public ToFunc
+{
+public:
+  //! \brief The constructor initializes the function to use.
+  GradientFunc(const FromFunc& f) : func(f) {}
+
+protected:
+  //! \brief Evaluates the gradient in a point.
+  Ret evaluate(const Vec3& X) const
+  {
+    return func.gradient(X);
+  }
+
+private:
+  const FromFunc& func; //!< Reference to scalar/vector function
+};
+
+
+//! \brief Specialization for symmetric tensors.
+//! \details We need to construct a symmtensor from the returned tensor.
+template<>
+SymmTensor GradientFunc<VecFunc,STensorFunc,SymmTensor>::evaluate(const Vec3& X) const
+{
+  Tensor tmp = func.gradient(X);
+  const size_t nsd = tmp.dim();
+  SymmTensor ret(nsd, nsd == 2 && tmp.size() == 4);
```

Comment:
This is a bit confusing. `func.gradient()` will here always return a standard tensor (2x2, size=4) or (3x3, size=9), afaik. So in the 2D case you are here constructing a symmetric tensor with `with33=true`, which is intended for axisymmetric stress tensors only. That does not look right. Maybe you intended to symmetrize the sensor instead, i.e., invoke the `symmetrize()` method?